lib/scholar/metrics/neighbors.ex

```diff
@@ -0,0 +1,35 @@
+defmodule Scholar.Metrics.Neighbors do
+  import Nx.Defn
+
+  deftransform recall(graph_true, graph_pred) do
+    if Nx.rank(graph_true) != 2 do
+      raise ArgumentError,
+            """
+            expected true graph to have shape {num_samples, num_neighbors}, \
+            got tensor with shape: #{inspect(Nx.shape(graph_true))}\
+            """
+    end
+    if Nx.rank(graph_pred) != 2 do
+      raise ArgumentError,
+            """
+            expected predicted graph to have shape {num_samples, num_neighbors}, \
+            got tensor with shape: #{inspect(Nx.shape(graph_pred))}\
+            """
+    end
+    if Nx.shape(graph_true) != Nx.shape(graph_pred) do
+      raise ArgumentError,
+            """
+            expected true and predicted graph to have the same shape, \
+            got #{inspect(Nx.shape(graph_true))} and #{inspect(Nx.shape(graph_pred))}\
+            """
+    end
+    recall_n(graph_true, graph_pred)
+  end
+
+  defn recall_n(graph_true, graph_pred) do
+    {n, k} = Nx.shape(graph_true)
+    concatenated = Nx.concatenate([graph_true, graph_pred], axis: 1) |> Nx.sort(axis: 1)
+    duplicate_mask = concatenated[[.., 0..(2 * k - 2)]] == concatenated[[.., 1..(2 * k - 1)]]
+    duplicate_mask |> Nx.sum() |> Nx.divide(n * k)
+  end
```

Comment:
Does this function calculate something different than recall from `Scholar.Metrics`?

Reply:
@msluszniak Merry Christmas!

I suppose you mean `Scholar.Metrics.Classification.recall/3`. Yes, this one is specific for algorithms that compute approximate k-nearest neighbors. It calculates the fraction of true nearest neighbors returned by the algorithm. In other words, it calculates how many nearest neighbors out of k possible did it get correctly, then averages over all the points. The meaning is similar to the one used for classification, but the way it is calculated is different (this one has to use sorting).

Reply:
Merry Christmas!
Ok, I see now :)